google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryAggregationsTest.java

```diff
@@ -1,5 +1,750 @@
+    writeAllDocs(writer, docs);
+    return collection;
+  }
+
+  public static void writeAllDocs(
+      CollectionReference collection, Map<String, Map<String, Object>> docs)
+      throws InterruptedException {
+    for (Map.Entry<String, Map<String, Object>> doc : docs.entrySet()) {
+      await(collection.document(doc.getKey()).set(doc.getValue()));
+    }
+  }
+
+  private static Map<String, Map<String, Object>> testDocs1 =
+      map(
+          "a",
+          map("author", "authorA", "title", "titleA", "pages", 100, "foo", 1, "bar", 2, "baz", 3),
+          "b",
+          map("author", "authorB", "title", "titleB", "pages", 50, "foo", 1, "bar", 2, "baz", 3));
+
+  @Test
+  public void canRunCountUsingAggregationMethod() throws Exception {
+    CollectionReference collection = testCollectionWithDocs(testDocs1);
+    AggregateQuerySnapshot snapshot = collection.aggregate(AggregateField.count()).get().get();
+    assertThat(snapshot.getCount()).isEqualTo(2);
+  }
+
+  @Test
+  public void canGetDuplicateAggregations() throws Exception {
+    CollectionReference collection = testCollectionWithDocs(testDocs1);
+    AggregateQuerySnapshot snapshot =
+        collection
+            .aggregate(AggregateField.count(), AggregateField.count(), sum("pages"), sum("pages"))
+            .get()
+            .get();
+    assertThat(snapshot.getCount()).isEqualTo(2);
+    assertThat(snapshot.get(sum("pages"))).isEqualTo(150);
+  }
+
+  @Test
+  public void canRunSumQuery() throws Exception {
+    CollectionReference collection = testCollectionWithDocs(testDocs1);
+    AggregateQuerySnapshot snapshot = collection.aggregate(sum("pages")).get().get();
+    assertThat(snapshot.get(sum("pages"))).isEqualTo(150);
+  }
+
+  @Test
+  public void canRunAverageQuery() throws Exception {
+    CollectionReference collection = testCollectionWithDocs(testDocs1);
+    AggregateQuerySnapshot snapshot = collection.aggregate(average("pages")).get().get();
+    assertThat(snapshot.get(average("pages"))).isEqualTo(75.0);
+  }
+
+  @Test
+  public void canGetMultipleAggregationsInTheSameQuery() throws Exception {
+    CollectionReference collection = testCollectionWithDocs(testDocs1);
+    AggregateQuerySnapshot snapshot =
+        collection.aggregate(sum("pages"), average("pages"), AggregateField.count()).get().get();
+    assertThat(snapshot.get(sum("pages"))).isEqualTo(150);
+    assertThat(snapshot.get(average("pages"))).isEqualTo(75.0);
+    assertThat(snapshot.get(AggregateField.count())).isEqualTo(2);
+  }
+
+  @Test
+  public void getCorrectTypeForSum() throws Exception {
+    CollectionReference collection = testCollectionWithDocs(testDocs1);
+    AggregateQuerySnapshot snapshot = collection.aggregate(sum("pages")).get().get();
+    Object sum = snapshot.get(sum("pages"));
+    assertThat(sum instanceof Long).isTrue();
+  }
+
+  @Test
+  public void getCorrectTypeForAverage() throws Exception {
```

Comment:
I like these two tests! Wonder if we should cover all possible return types? Sum: Double, Long, NaN. Average: Double, NaN, Null

Reply:
Done